### PR TITLE
Domains: checkout/whois form consolidation

### PR DESCRIPTION
**Pull request by ramonjd**
This PR addresses issue #18808

## Goals
The core objective is to consolidate the domain checkout and whois contact details form, which share common fields and functionality.

Because of the varying internal business logic of each form, there's no way to unify them completely, however we can attempt to reduce some of the repetition in the contact form itself. 

## First steps
- [x] identify common functionality
- [x] create a `<ContactDetailsFormFields />` common component
- [x] migrate formState controller into `<ContactDetailsFormFields />`
- [x] integration with domain details form
- [x] integrate with WHOIS contact details form 

## TODO
Update the button class attribute value for the [wp-e2e-test](https://github.com/Automattic/wp-e2e-tests/blob/master/lib/pages/signup/checkout-page.js).

### Domain checkout form
checkout-after

### WHOIS form
whois-after

## Afterthoughts
This was a worthy exercise, not only because it exposed the idiosyncratic logic and dependencies of both form, but also highlighted areas of complexity, not limited to the following:

- Dealing with the specific logic of each journey, for example, disabling form fields and rendering a 'GApps' state, led to a component that only really shifts the work to another place.
- `form-state` triggers a great deal of rendering and is a little wily

Furthermore, the effort required to consolidate the forms has been quite high, and might turn out to be disproportionate to the need for simplifying two contact forms.

As mentioned, the two forms perform separate tasks, though they share the same contact details requirements. Creating a common dependency could create more work should their internal logic diverge further later. Rather than instructing a swiss-army-like `<ContactDetailsFormFields />` to render a form permutations only it knows about, we might come to the conclusion that it would be better to instruct a form-builder component to construct a form _using fields controls we specify_.

### CODE REVIEW

File: client/components/contact-details-form-fields/index.jsx [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+	renderContactDetailsFields() {
+		const { translate, needsFax, hasCountryStates } = this.props;
+		const countryCode = this.getCountryCode();
+		const { phoneCountryCode } = this.state;
+
+		return (
+			<div className="checkout__domain-details-contact-details-fields">
+				{ this.createField( 'organization', HiddenInput, {
+					label: translate( 'Organization' ),
+					text: translate( "+ Add your organization's name" ),
```
Comment (a8ci18n) [inline client/components/contact-details-form-fields/index.jsx]: Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Add Organization Name' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/contact-details-form-fields/index.jsx [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+	renderContactDetailsFields() {
+		const { translate, needsFax, hasCountryStates } = this.props;
+		const countryCode = this.getCountryCode();
+		const { phoneCountryCode } = this.state;
+
+		return (
+			<div className="contact-details-form-fields__container">
+				{ this.createField( 'organization', HiddenInput, {
+					label: translate( 'Organization' ),
+					text: translate( "+ Add your organization's name" ),
```
Comment (a8ci18n) [inline client/components/contact-details-form-fields/index.jsx]: Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Add Organization Name' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/contact-details-form-fields/index.jsx [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+		const { phoneCountryCode } = this.state;
+
+		return (
+			<div className="contact-details-form-fields__contact-details">
+				{ this.createField(
+					'organization',
+					HiddenInput,
+					{
+						label: translate( 'Organization' ),
+						text: labelTexts.organization || translate( '+ Add organization name' ),
```
Comment (a8ci18n) [inline client/components/contact-details-form-fields/index.jsx]: Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Add Organization Name' )` *ES Score: 17*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/checkout-whois-form-consolidation)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/domains/contact-details-form-fields/index.jsx [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+		const { phoneCountryCode } = this.state;
+
+		return (
+			<div className="contact-details-form-fields__contact-details">
+				{ this.createField(
+					'organization',
+					HiddenInput,
+					{
+						label: translate( 'Organization' ),
+						text: labelTexts.organization || translate( '+ Add organization name' ),
```
Comment (a8ci18n) [inline client/components/domains/contact-details-form-fields/index.jsx]: Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Add Organization Name' )` *ES Score: 17*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/checkout-whois-form-consolidation)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/phone-input/index.jsx [hunk 5]
```diff
@@ -243,7 +243,8 @@ class PhoneInput extends React.PureComponent {
 					name={ this.props.name }
 					ref={ this.setNumberInputRef }
 					type="tel"
-					className={ classnames( { 'is-error': this.props.isError } ) }
+					disabled={ this.props.disabled }
+					className={ this.props.isError && 'is-error' }
```
Comment (klimeryk) [inline client/components/phone-input/index.jsx]: It seems that once there's a validation error in the phone field, it stays in the `is-error` state 🤔

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 6]
```diff
@@ -344,117 +156,46 @@ export class DomainDetailsForm extends PureComponent {
 	renderPrivacySection() {
 		return (
 			<PrivacyProtection
-				checkPrivacyRadio={ this.allDomainItemsHavePrivacy() }
```
Comment (klimeryk) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: It seems the `PrivacyProtection` component was not updated after this rename and it's still relying on `this.props.checkPrivacyRadio`, which leads to this interesting state:
screen shot 2017-12-03 at 11 09 36

File: client/my-sites/checkout/checkout/domain-details-form.jsx [hunk 7]
```diff
@@ -318,11 +134,8 @@ export class DomainDetailsForm extends PureComponent {
 		return cartItems.hasOnlyDomainRegistrationsWithPrivacySupport( this.props.cart );
 	}
 
-	allDomainItemsHavePrivacy() {
-		return (
-			cartItems.getDomainRegistrationsWithoutPrivacy( this.props.cart ).length === 0 &&
-			cartItems.getDomainTransfersWithoutPrivacy( this.props.cart ).length === 0
-		);
+	allDomainRegistrationsHavePrivacy() {
```
Comment (klimeryk) [inline client/my-sites/checkout/checkout/domain-details-form.jsx]: Why only domain registrations? Domain transfers are domains too! (or was this discussed with @aidvu?)

Comment (aidvu) [reply client/my-sites/checkout/checkout/domain-details-form.jsx]: This is probably why it needs a rebase. :)

Comment (klimeryk) [reply client/my-sites/checkout/checkout/domain-details-form.jsx]: Aaaah, damn GitHub, confusing me! 🤜 :octocat:  🤛

Comment (ramonjd) [reply client/my-sites/checkout/checkout/domain-details-form.jsx]: Going to rebase as the first task 👍

File: client/components/domains/contact-details-form-fields/index.jsx [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+			loadFunction: this.loadFormState,
+			sanitizerFunction: this.sanitize,
+			validatorFunction: this.validate,
+			onNewState: this.setFormState,
+			onError: this.handleFormControllerError,
+			debounceWait: 500,
+		} );
+	}
+
+	loadFormState = fn => {
```
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/components/domains/contact-details-form-fields/index.jsx]: It's legacy from the previous `form-state` hydration. You're right, it's a little opaque. Will change.

File: client/components/domains/contact-details-form-fields/README.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+Usage:
+
+```js
+    // the method to be passed as a prop to <ContactDetailsFormFields />
+	onContactDetailsChange = ( newContactDetails ) => {
+		this.props.updateReduxStore( newContactDetails );
+	};
+```
+
+### onContactDetailsChange {Func} (required)
```
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/components/domains/contact-details-form-fields/README.md]: Hawkeye!
Now that I'm reviewing it, I've actually noticed several mistakes in there ;)
Thanks!

File: client/components/domains/contact-details-form-fields/index.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (klimeryk) [reply client/components/domains/contact-details-form-fields/index.jsx]: Ah, should have noticed that, sorry about that. But, yeah, let's rename it to something more descriptive 🙇

File: client/my-sites/checkout/checkout/domain-details-form.jsx (hunk 6 above)
Comment (ramonjd) [reply client/my-sites/checkout/checkout/domain-details-form.jsx]: How can I reproduce this? 

FYI, I ditched most of the props on this component since `countriesList`, `disabled`, `fields`, and `onCheckboxChange` weren't being used by `<PrivacyProtection />` at all. 

And `this.handleCheckboxChange` didn't even exist in the domain details component :)

**Comment by ramonjd** [comment]
@klimeryk Huge thanks for your review so far. Really helpful.

> Apart from what I've mentioned in the comments, I've noticed that the check introduced in #16240 was undone by this - we should not allow the user to re-save the same WHOIS info.

I had this on my radar, and I thought it was working. I must have broken my fix. :)

File: client/components/phone-input/index.jsx (hunk 5 above)
Comment (ramonjd) [reply client/components/phone-input/index.jsx]: Definitely 🤔 ... Haven't been able to reproduce this yet. But I'm going to try!

File: client/components/domains/contact-details-form-fields/index.jsx [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+import RegionAddressFieldsets from './custom-form-fieldsets/region-address-fieldsets';
+import notices from 'notices';
+import support from 'lib/url/support';
+
+const countriesList = countriesListForDomainRegistrations();
+
+export class ContactDetailsFormFields extends Component {
+	static propTypes = {
+		eventFormName: PropTypes.string,
+		contactDetails: PropTypes.shape( {
```
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/components/domains/contact-details-form-fields/index.jsx]: I like less repetition :) It makes sense, since they are, and must be, equal. 

We'd have to pull `this.fieldNames` out of the constructor. Making it another static var would work (and we could access it through `ContactDetailsFormFields.fieldNames`).

But now I reflect on it, `this.fieldNames` is a definitive and immutable list of form fields. It might work better as a constant. 👍

File: client/components/phone-input/index.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/components/phone-input/index.jsx]: Rebase fixed it! :)

**Comment by ramonjd** [comment]
> However, when you hit Request Confirmation button, the main area of Calypso goes blank and there's this confusing error in the console

My mistake. I wasn't caching the form values. Also noticed the effect of #16239 **Domains: After updating, the WHOIS store is not automatically updated**, whereby the button isn't disabled. I might take a look at that later, but I think we just need to do another `upgradesActions.fetchWhois` to update the store after submit, or something... :)

**Comment by klimeryk** [comment]
> I was deleting the fax (as we were doing in the original domain-details-form.jsx) and therefore the comparison check failed. 👍

That probably worked because we use Tucows for new domains, while this empty value seems to be required by WWD only 🤷‍♂️

File: client/components/domains/contact-details-form-fields/index.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (klimeryk) [reply client/components/domains/contact-details-form-fields/index.jsx]: Ah, right, I did not think about the static/instance problem, but yup, you're right - it's a constant/immutable list anyway.

File: client/my-sites/checkout/checkout/domain-details-form.jsx (hunk 6 above)
Comment (klimeryk) [reply client/my-sites/checkout/checkout/domain-details-form.jsx]: IIRC, I did not do anything special - just try to register a new domain that supports privacy. And try to check a different radio button - both will be checked as a result.

**Comment by ramonjd** [comment]
Merged in 05152d3 from #20650

**Comment by ramonjd** [comment]
Rebase and hooked in PR #20549

**Comment by klimeryk** [comment]
Sorry for letting this one sit around so long, @ramonjd - it's #1 on my review list for tomorrow 🙇

**Comment by ramonjd** [comment]
No stress @klimeryk 

Actually, it's given us time to reflect on how useful/limiting `formState` is here in this instance; basically where we should do the 'controlling' of the controlled form. I've been chatting with @deBhal about it, and what I've done in this PR is incompatible with #21226 

Primarily how we're passing down props bound to the `formState` in `domain-details-form`. 

We've been mulling over a patch for it, but it's one of those _"We can put a plaster on it, or amputate your arm, Mrs Smith"_ :)

**Comment by aidvu** [comment]
Oooops... I forgot this, again... 🤦‍♂️

**Comment by klimeryk** [comment]
@ramonjd @deBhal Understood - but I'd prefer to first consolidate those two forms and then refactor how data flows through them. What do you think?

**Comment by ramonjd** [comment]
@deBhal do you think we can get those forms working as discussed without passing down the bound prop method?

It would be good to kick this PR into goal. Then we can come back and slap `formState` into shape somehow (?)

**Comment by deBhal** [comment]
We're looking at addressing the conflict in #21781, and then we can circle back to this.

**Comment by aidvu** [comment]
Thanks for the update @deBhal.

**Comment by ramonjd** [comment]
> We're looking at addressing the conflict in #21781, and then we can circle back to this.
Fixed in #22156. Full steam ahead 🚂 !!

**Comment by ramonjd** [comment]
Here are some bundle stats for this branch. 

**TL;DR** This PR adds circa 20k to the gzipped domains chunk mainly due to phone number validation code. 

----

Looks like it's adding about 20k to the gzipped payload in domains/components. From what I can tell, more than 3/4 of this comes down to `<FormPhoneMediaInput />` and its dependency [data.js](https://github.com/Automattic/wp-calypso/blob/c8142f9/client/components/phone-input/data.js
), which contains the rules for validating phone numbers. Both comprise a fairly hefty slab of code that's only used in this form.

phone-input screen shot 2018-02-08 at 12 13 46 pm

Previously we were only pulling this component in the __checkout__ chunk, but with the consolidated form it's now in __domains__.

I experimented with loading the shared `<ContactDetailsFormFields />` using `<AsyncLoad />`. This reduces the payload overall, only slightly, but creates a 25k chunk that has to be downloaded. Given that checkout is one of our most important flows, I'd argue we wouldn't want to risk keeping the user waiting for a sequential HTTP call. 

with async screen shot 2018-02-08 at 12 11 37 pm

Instead, I'm inclined to see if we can load [data.js](https://github.com/Automattic/wp-calypso/blob/c8142f9/client/components/phone-input/data.js
) asynchronously, and refactor the phone input to handle the case when no validation is present. I guess this depends on how crucial phone validation is to WordPress / the domain registrar.

Overall, it's a slight step backwards in the quest to lower our bundle, but we might be able to claw back some of those bytes later. (?)

File: client/components/domains/contact-details-form-fields/index.jsx [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+		labelTexts: {},
+		onCancel: null,
+		disableSubmitButton: false,
+		className: '',
+		needsOnlyGoogleAppsDetails: false,
+		hasCountryStates: false,
+		translate: identity,
+	};
+
+	constructor() {
```
[the remark this answers is not in the document]
Comment (klimeryk) [reply client/components/domains/contact-details-form-fields/index.jsx]: Still no `props`? 😢

**Comment by ramonjd** [comment]
Thanks @klimeryk !
Wouldn't have got this one even near the goal posts without you and everyone's reviews! I'm going to look at how to reduce the phone `data.js` when I get time.

> Still no props?

Do you mean the `props` arg in the constructor? We don't need to since we're not accessing `this.props` in the constructor, so it's fine to leave out.

> please merge this when there are many HEs around, just in case there's a regression, so that it will be quickly reported and handled appropriately

Will do. 🙇

**Comment by klimeryk** [comment]
> Do you mean the props arg in the constructor? We don't need to since we're not accessing this.props in the constructor, so it's fine to leave out.

That might be true, but in the future it might change and someone will get unpleasantly surprised ;) You're calling the parent constructor either way, might as well pass the props to it ¯\\\_(ツ)\_/¯

**Comment by klimeryk** [comment]
Please merge this, @ramonjd, before it grows any more merge conflicts 😅 🌵

**Comment by alisterscott** [comment]
Hi 👋

I noticed a couple of potential issues with this:

Firstly, at ~900px or narrower the state and postcode fields merge into one. Also, the "Continue to Checkout" label has been renamed to a less descriptive "Continue" - was this intentional? I didn't see this mentioned in the PR.

I'll raise the fields issue as a separate bug

**Now**

now

**Previously**

previous

**Comment by ramonjd** [comment]
@alisterscott Thanks for testing! Appreciate it. And for catching the classname change in the e2e tests. Totally slipped past me.

I'll look into those issues today.
